lambda/matching.ml

```diff
@@ -1028,6 +1028,15 @@ let is_or p =
   | Tpat_or _ -> true
   | _ -> false
 
+let rec heads_are_var p =
+  match p.pat_desc with
+  | Tpat_any
+  | Tpat_var _ ->
+      true
+  | Tpat_alias (p, _, _) -> heads_are_var p
+  | Tpat_or (p1, p2, _) -> heads_are_var p1 && heads_are_var p2
```

Comment:
Is it necessary to check `p2` here? After all, when simplifying `Tpat_or`, we replace `ω | …` by `ω` directly. It seems that we could implicitly apply this rule here and only have `heads_are_var p1`.

Reply:
That makes sense to me, but given it's @gasche's patch, let's see what he thinks.
(I personally suggested replacing `Tpat_or (p1, p2, _)` by `Tpat_or (p1, p2, None)`, since in the `Some _` case we know the heads are polymorphic variant constructors; but @gasche didn't seem to like that "optimization")

Reply:
I don't think it is wise to change the code based on what we know simplification is going to do in the future -- it makes the code harder to understand by adding a dependency on invariants from some other part of the code (and: we discussed disabling this simplification anyway, assuming that it is uncommon as it fires the "unused subpattern" warning). The logic of that particular function is that it is a specialized instance of a generic "collect all the potential heads of this pattern" followed by an implicit `List.for_all` traversal; in general the heads of `p | q` are the heads of `p` and the heads of `q`, and following this general pattern gives a clear structure to the code.

Reply:
After more in-person grumbling from @Octachron we agreed to change the definition to use `... p1 || ... p2`, and I renamed the function (jamais deux sans trois) into `omega_like`.